### PR TITLE
feat: BALANCE Opcode

### LINKED ISSUE #31: Opcode: BALANCE

### PR DESCRIPTION
**Pull request by FrancoGiachetta**
Closes #31 
This PR handles the implementation of BALANCE opcode

### CODE REVIEW

**Comment by FrancoGiachetta** [comment]
Don't really know if the attributes of DbAccounts were private on purpose. I made them public so that I could make the tests.

File: tests/evm.rs [hunk 1]
```diff
@@ -1,5 +1,9 @@
+use std::collections::HashMap;
+
+use ethereum_types::Address;
```
Comment (JulianVentura) [inline tests/evm.rs:3]: Please, import types from `src/primitives.rs`

File: tests/evm.rs [hunk 2]
```diff
@@ -454,3 +458,80 @@ fn callvalue_stack_overflow() {
...
+    let env = Env::default();
+
+    run_program_assert_result(operations, env, BigUint::from(0_u8));
+}
+
+#[test]
+fn balance_with_existing_account() {
+    let address = Address::from_low_u64_be(123456);
+    let balance = ethereum_types::U256::from_dec_str("123456").unwrap();
+    dbg!("Balance2 {}", balance);
```
Comment (JulianVentura) [inline tests/evm.rs:477]: Here there is another debug log

File: tests/evm.rs [hunk 3]
```diff
@@ -454,3 +458,80 @@ fn callvalue_stack_overflow() {
...
+        Operation::Balance,
+    ];
+    let env = Env::default();
+
+    run_program_assert_result(operations, env, BigUint::from(0_u8));
+}
+
+#[test]
+fn balance_with_existing_account() {
+    let address = Address::from_low_u64_be(123456);
```
Comment (JulianVentura) [inline tests/evm.rs:475]: Can you please modify this test to match the [example of the EVM Codes](https://www.evm.codes/playground?callValue=125985&unit=Wei&codeType=Mnemonic&code=%27%2F%2F%20Read%20current%20contract%20balance%5CnADDRESS%5CnBALANCE%27_)? You can look at [CALLER](https://github.com/lambdaclass/evm_mlir/pull/260) or [ADDRESS](https://github.com/lambdaclass/evm_mlir/pull/254) PRs to see an example.

File: README.md [hunk 4]
```diff
@@ -39,10 +39,12 @@ An EVM-bytecode to machine-bytecode compiler using MLIR and LLVM.
 1. (0x1B) SHL
 1. (0x1C) SHR
 1. (0x1D) SAR
+1. (0x31) BALANCE
 1. (0x34) CALLVALUE
 1. (0x35) CALLDATALOAD
 1. (0x36) CALLDATASIZE
 1. (0x38) CODESIZE
+1. (0x43) NUMBER
```
[the remark this answers is not in the document]
Comment (FrancoGiachetta) [reply README.md:47]: This thing is that a realized a forgot to change it on that PR.

[the remark this answers is not in the document]
Comment (JulianVentura) [reply README.md:47]: Ok

File: src/codegen/context.rs [hunk 5]
```diff
@@ -677,4 +677,22 @@ impl<'c> OperationCtx<'c> {
             location,
         )
     }
+
+    #[allow(clippy::too_many_arguments)]
```
Comment (MegaRedHand) [inline src/codegen/context.rs:681]: I don't think this is necessary in this case.

File: src/codegen/operations.rs [hunk 6]
```diff
@@ -1875,6 +1876,93 @@ fn codegen_sar<'c, 'r>(
...
+
+    let address = stack_pop(context, &ok_block)?;
+
+    let address_ptr = ok_block
+        .append_operation(llvm::alloca(
+            context,
+            pointer_size,
+            ptr_type,
+            location,
+            AllocaOptions::new().elem_type(Some(TypeAttribute::new(h160.into()))),
```
Comment (MegaRedHand) [inline src/codegen/operations.rs:1922]: Let's do the alloca for an uint256 instead.

File: src/syscall.rs [hunk 7]
```diff
@@ -310,6 +314,27 @@ impl<'c> SyscallContext<'c> {
...
         self.env.tx.data.as_ptr()
     }
+
+    pub extern "C" fn get_balance(&mut self, address: &U256, balance: &mut U256) {
+        let address_hi_slice = address.hi.to_be_bytes();
+        let address_lo_slice = address.lo.to_be_bytes();
+        let address_slice = [&address_hi_slice[12..16], &address_lo_slice[..]].concat();
+
+        let address = Address::from_slice(&address_slice);
+        dbg!("{}", &self.db);
```
Comment (MegaRedHand) [inline src/syscall.rs:324]: We can remove this now.

File: src/syscall.rs [hunk 8]
```diff
@@ -310,6 +314,27 @@ impl<'c> SyscallContext<'c> {
...
+        match self.db.basic(address).unwrap() {
+            Some(a) => {
+                balance.hi = (a.balance >> 128).low_u128();
+                balance.lo = a.balance.low_u128();
+                dbg!("SOME");
+            }
+            None => {
+                balance.hi = 0;
+                balance.lo = 0;
+                dbg!("NONE");
```
Comment (MegaRedHand) [inline src/syscall.rs:334]: Same here

File: tests/evm.rs [hunk 9]
```diff
@@ -650,3 +650,71 @@ fn chainid_stack_overflow() {
     let env = Env::default();
     run_program_assert_halt(program, env);
 }
+
+#[test]
+fn balance_with_unexisting_account() {
```
Comment (MegaRedHand) [inline tests/evm.rs:655]:
```suggestion
fn balance_with_non_existent_account() {
```

File: src/syscall.rs [hunk 10]
```diff
@@ -310,6 +314,27 @@ impl<'c> SyscallContext<'c> {
     pub extern "C" fn get_calldata_ptr(&mut self) -> *const u8 {
         self.env.tx.data.as_ptr()
     }
+
+    pub extern "C" fn get_balance(&mut self, address: &U256, balance: &mut U256) {
```
Comment (MegaRedHand) [inline src/syscall.rs:318]: Let's rename it to `store_in_...` like the rest

File: tests/evm.rs (hunk 9 above)
Comment (MegaRedHand) [reply tests/evm.rs:655]: Also, let's add another test for addresses that use the whole 256 bits (which are obviously invalid/non-existent).

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

**Comment by ricomateo** [comment]
LGTM @MegaRedHand

**Review by MegaRedHand — APPROVED** [review]
LGTM

**Pull request merged**